### ISSUE TITLE
SIngle Container version of Compreface

### ISSUE THREAD

**Issue by alexdelprete** [exadel-inc/CompreFace#651]
**Is your feature request related to a problem? Please describe.**
We're using Compreface as the recog engine of Double Take (an Home Assistant addon): https://github.com/jakowenko/double-take

Double Take dev (@jakowenko) created the addons for Double Take and other recognition frameworks: DeepStack and Facebox, but our preferred solution is Compreface. Unfortunately we can only build an addon starting from a single container, while Compreface requires five separate containers, unlike DeepStack and Facebox.

**Describe the solution you'd like**
A single-container build of Compreface, so we can create an Home Assistant addon for Double Take users, that right now are using DeepStack only because there's an addon available, while we recommend Compreface for best recognition results.

**Comment by pospielov** [comment]
I don't think that running CompreFace in one image is a good idea.
Now if something wrong - you can just restart one of the services, while others continue to work.
But I got the idea - it should make it much easier to run with Home Assistant.
I'll add it to our backlog and try to make it as soon as we have resources for it.

I haven't used HA, but I see that it has another thing - integrations. In this case, you have to run CompreFace himself, but there is still an integration with HA. Will it work for you?

**Comment by alexdelprete** [comment]
I agree with you, but in order to use Compreface as an addon in HassOS, we need a single container. And I'm not asking this just for myself, there are literally hundreds of users that are using Double Take + Frigate to do face recognition automation, and unfortunately they are forced to use DeepStack now because it's the only addon available, while advanced users are installing Compreface manually, even if HA doesn't support it, because of its superior recognition capabilities.

We are recommending users to use Compreface but the feedback is that it's too hard to install. So we need to develop the addon to install it directly in HA, like it has been done for DeepStack. It's a pity these users have to revert to an inferior solution like DeepStack, they deserve Compreface, I hope you can understand this and help us reach the objective.

As you can see here, DoubleTake is the only addon that @jakowenko has been able to implement, we are waiting for a single container version of Compreface now so users can have the better choice.

![image](https://user-images.githubusercontent.com/7027842/139906265-7cc4e8c0-641b-45be-8cd0-4a63968f8158.png)

> I haven't used HA, but I see that it has another thing - integrations. In this case, you have to run CompreFace himself, but there is still an integration with HA. Will it work for you?

Compreface addon will be used as an engine by the Double Take addon, right now we are using Compreface in an unsupported way manually installing it into HA's docker. Once installed, Double Take gets the snapshots from Frigate, and passes them to Compreface for recognition, then it processes the result and creates HA sensors to do the automations. It's working beautifully but this kind of setup is unsupported unfortunately so users are reverting to DeepStack. :(

Thanks for your reply and I hope you can deliver the single container version as soon as possible. :)

**Comment by bigbangus** [comment]
I second this also for the Unraid community. It would be ideal to have it in a single container that can pulled from the App store template. Currently it's a steep learning curve and requires the "compose.manager" plugin to work. So the majority of users will default to deepstack with reportedly poor recognition performance when pairing frigate with double take for facial recognition on rtsp feeds.

**Comment by alexdelprete** [comment]
> So the majority of users will default to deepstack with reportedly poor recognition performance when pairing frigate with double take for facial recognition on rtsp feeds.

That's the main problem we're facing now for the [Double Take project](https://github.com/jakowenko/double-take): 90% of support questions are for Deepstack installation and poor recognition rate.

We need to switch them to Compreface, but we need a single container version to be able to do that.

**Comment by pospielov** [comment]
I am working on it right now and already have the first working version. There are still some problems. In an optimistic scenario, I'll publish it this week and write a short description of how to run it here. Normal readme I'll write a little bit later.

**Comment by alexdelprete** [comment]
> I am working on it right now and already have the first working version. There are still some problems. In an optimistic scenario, I'll publish it this week and write a short description of how to run it here

Amazing news Pospielov. Thank you so much. Users will be happy to switch from DeepStack to CompreFace, like we suggested.

Please let me know when it's ready so we can test it. Thank you.

**Comment by bigbangus** [comment]
> I am working on it right now and already have the first working version. There are still some problems. In an optimistic scenario, I'll publish it this week and write a short description of how to run it here. Normal readme I'll write a little bit later.  

Thank you. A lot of Unraid users have an Nvidia GPU for containers like plex so if you could include a method to pass extra parameters for the GPU custom builds also that would be ideal. Again thank you for your effort and time.

**Comment by Iceman248** [comment]
Thanks for the work on this. We greatly appreciate it and look forward to the end result.

**Comment by corgan2222** [comment]
> 
> 
> I am working on it right now and already have the first working version. There are still some problems. In an optimistic scenario, I'll publish it this week and write a short description of how to run it here. Normal readme I'll write a little bit later.

Awesome, that's great News!

I have written the[ unraid docker templates for double-take](https://forums.unraid.net/topic/114590-support-corgans-docker-repository-double-take-facebox-tagbox-videobox) and facebox, but because [unraid did not support docker-compose](https://forums.unraid.net/topic/114598-how-to-setup-an-docker-unraid-template-from-docker-compose) the users are stuck to a command-line installation, which isn't ideal for the average user.

As @bigbangus mentioned, if you could add support for GPU processing, it would be awesome, but not mandatory. 
So just comment here, or hit me directly, if you have something to test. 
I would then transfer your docker stuff into an unraid app template for the Community Store.

**Comment by pospielov** [comment]
Hi all, it looks like I managed to make a single image of CompreFace
How to run:
`docker run -it --name=CompreFace -v compreface-db:/var/lib/postgresql/data -p 8000:80 exadel/compreface:0.6.1`
where:
`name=CompreFace` - is the name of container
`compreface-db` - is the name of the volume. This is important to keep the data if you delete the container.
Under the hood is used `supervisord`. It helps to achieve similar behavior with docker-compose - if the service fails, it automatically starts it.
Some other notices:
1. Compared to docker-compose we can't guarantee that DB starts before servers. I added 10 seconds of timeout between the start of DB and the first service. 
2. If DB starts too slow - service will fail. But `supervisord` will restart it. So in the bad scenario, you will see lots of errors in logs, but it will work
3. In the worst scenario it one of the services won't start, but `supervisord` will endlessly try to start it. You will see the same error in the console. In this case, write the log to me. 
4. Single container version of CompreFace isn't tested by a standard QA process. On another hand, I reuse the images from DockerHub. So basically the only difference is with the deploy method.
5. As I reuse the images from DockerHub, I managed to make all official versions of Compreface: FaceNet(default), Arcface-R100, Arcface-R100-GPU, MobileNet, MobileNet-GPU
To run the GPU version, you need to add `--runtime=nvidia`:
`docker run -it --name=CompreFace -v compreface-db:/var/lib/postgresql/data --runtime=nvidia -p 8000:80 exadel/compreface:0.6.1-arcface-r100-gpu`
6. The environment variables are still actual from `docker-compose` version, so to set API server limit you can run:
`docker run -it --name=test -e "API_JAVA_OPTS=-Xmx8g" -v compreface-test-db:/var/lib/postgresql/data --runtime=nvidia -p 8000:80 exadel/compreface:0.6.1-mobilenet-gpu`

**Comment by alexdelprete** [comment]
Pospielov, thank you so much, we will test and let you know in case of problems. @jakowenko will try to build an addon for Home Assistant out of this.

**Comment by Iceman248** [comment]
Thanks. I'll test once an unRAID app template is done. I look forward to that. 😀

**Comment by corgan2222** [comment]
Awesome! Will test this on Unraid!

**Comment by corgan2222** [comment]
@pospielov 

Thanks for this container! Runs great on Unraid with default FaceNet and arcface-r100-gpu.
But I have problems with the volume.

If I start the container without the volume, the /var/lib/postgresql/data Folder has the correct permission.
![grafik](https://user-images.githubusercontent.com/12233951/142712242-002324ed-2211-4d4a-b205-68e9af9b7a50.png)

But if I configure the volume path, with 

Unraid Docker Command:
`
/usr/local/emhttp/plugins/dynamix.docker.manager/scripts/docker create --name='CompreFace' --net='bridge' --privileged=true -e TZ="Europe/Berlin" -e HOST_OS="Unraid" -p '8800:80/tcp' -v '/mnt/user/appdata/compreface':'/var/lib/postgresql/data':'rw' --runtime=nvidia 'exadel/compreface:0.6.1-arcface-r100-gpu'`

I get these errors:

> 
> 2021-11-20 04:04:05,942 INFO success: compreface-postgres-db entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
> today at 04:04:05 2021-11-20 03:04:05.953 GMT [14] LOG:  skipping missing configuration file "/var/lib/postgresql/data/postgresql.auto.conf"
> today at 04:04:05 2021-11-20 03:04:05.953 UTC [14] FATAL:  "/var/lib/postgresql/data" is not a valid data directory
> today at 04:04:05 2021-11-20 03:04:05.953 UTC [14] DETAIL:  File "/var/lib/postgresql/data/PG_VERSION" is missing.
> 
> 021-11-20 03:06:59.579 GMT [45] LOG:  skipping missing configuration file "/var/lib/postgresql/data/postgresql.auto.conf"
> today at 04:06:59 2021-11-20 03:06:59.579 UTC [45] FATAL:  data directory "/var/lib/postgresql/data" has invalid permissions
> today at 04:06:59 2021-11-20 03:06:59.579 UTC [45] DETAIL:  Permissions should be u=rwx (0700) or u=rwx,g=rx (0750).
> 

and the permissions are set to 777.
Strange is, if i exec into the Container and set chmod to 755 I still get the same errors.

I tested the Container on several Linux machines, and it runs fine. So that must be a problem with the unraid docker system.
Any Ideas?

So ATM the Container is usable, but without a persistent data Folder.

**Comment by pospielov** [comment]
@corgan2222 in my example I used [named volume](https://docs.docker.com/storage/volumes/):
`-v compreface-db:/var/lib/postgresql/data`
The default behavior of named volume is that when it first created the volume, docker copies all content of the image to the volume with all permissions. Docker in its documentation recommends using named volumes because of portability.

When you set the exact folder to mount, docker created a [bind mounts](https://docs.docker.com/storage/bind-mounts/):
`-v '/mnt/user/appdata/compreface':'/var/lib/postgresql/data':'rw'`
By default, docker does not copy image content into bind mounts. This is why the Postgres data folder is empty inside your container and Postgres can't start. 

So I would follow docker recommendations and use named volumes. In case this is impossible or you still prefer using binding to a folder, there is a [workaround](https://stackoverflow.com/a/47528568/14485439) for such case if you still want to mount the exact folder into the docker container:
You need to create a `named bind mount`:
```
docker volume create --driver local \
    --opt type=none \
    --opt device=/mnt/user/appdata/compreface \
    --opt o=bind \
    compreface-db
```
And then use it as named volume during start:
`docker run -it --name=CompreFace -v compreface-db:/var/lib/postgresql/data -p 8000:80 exadel/compreface:0.6.1`

**Comment by corgan2222** [comment]
@pospielov  Thanks for clarification, now the behavior make sense.
The main problem is, that unraid user can't change the behavior of the docker process works.
This is how all docker volumes in unraid looks:

![grafik](https://user-images.githubusercontent.com/12233951/142779474-e9b89be8-3cd1-4e29-aaa1-c1f50ac19d21.png)

We as Apps Maintainer create only XML Templates. The Unraid GUI then creates the Docker Containers.
[This is how the Compreface XML looks.](https://github.com/corgan2222/unraid-templates/blob/main/CompreFace/CompreFace.xml)

GUI View:

![grafik](https://user-images.githubusercontent.com/12233951/142779538-d948bae6-ecd7-43f9-8b34-f8f224f66564.png)

I'm not that deep into the Unraid Docker System. Do you have an Idea left, or should I forward this to the Unraid DEvs?

**Comment by pospielov** [comment]
What if you in `Host Path` put value without `/` at the beginning? Like `compreface-db`
If it won't work, then probably this is intentional.
https://forums.unraid.net/topic/88920-named-volumes-vs-bind-mount/
I see that they don't recommend named volumes. I would like to put them with docker developers in one room and they argue with each other :)
So, if it won't work, this is a good idea to ask Unraid developers what is their best practices in this case.
I see one workaround - I can put a backup of all empty postgres data files somewhere and during startup check, if the `/var/lib/postgresql/data` folder is empty, then copy the backup files into it. But this looks like a hack, not a normal implementation.

**Comment by bigbangus** [comment]
Not sure if this helps regarding mounted volumes, but I'm running compreface now in Unraid with the modifications to docker-compose.yml below and it works fine. (not using app templates, but docker compose plugin)

@pospielov  are you after performance when you decided to use a named volume instead of a bind mount?
I found this thread that supports this idea: https://stackoverflow.com/questions/64629569/docker-bind-mount-directory-vs-named-volume-performance-comparison
So maybe in Windows/MacOS, compreface would be quicker with named volumes? I don't know. It just seems that every docker container on Unraid (of which there are thousands) all use bind mounts to separate application from data.

```
version: '3.4'

###volumes:
###  postgres-data:

services:
  compreface-postgres-db:
    restart: always
    image: postgres:11.5
    restart: always
    container_name: "compreface-postgres-db"
    environment:
      - POSTGRES_USER=${postgres_username}
      - POSTGRES_PASSWORD=${postgres_password}
      - POSTGRES_DB=${postgres_db}
    volumes:
###      - postgres-data:/var/lib/postgresql/data
      - /mnt/user/appdata/compreface/postgres:/var/lib/postgresql/data
```

![image](https://user-images.githubusercontent.com/91270101/142900026-87062774-7658-4325-a26a-78e4619e2d51.png)

**Comment by pospielov** [comment]
@bigbangus CompreFace uses named volumes by default because this is the only way it can work for everyone. 
I mean bind mounts require the path and it depends on the folder structure on our machine and the operations system.
So the only way to guarantee volume creating is using named volumes.

Of course, experienced users can change it if they want. 
I don't expect serious performance issues during recognition with Compreface and bind mounts if you use double-take. 
We cache all required data for recognition in the memory. So as double-take always uses one api-key, you can be sure that CompreFace won't read from the database each recognition request.
Of course, when you save a new example, CompreFace will save it to the database and sync the cache. So this may be a slow point. But I don't think this is a common operation.

Also, your docker-compose file adds even more questions. So you also use bind mounts. But in your case docker copied image content into the folder. I don't know why - probably docker-compose does something similar to what I showed as a workaround. Still, as I understand, the problem is that Unraid doesn't support it.

**Comment by corgan2222** [comment]
> I don't know why - probably docker-compose does something similar to what I showed as a workaround. Still, as I understand, the problem is that Unraid doesn't support it.

Yes, that's also for me the main point. 
I will look into this topic next weekend, as I'm on a working trip this week. Thanks for your help so far!

@bigbangus: the reason for this ticket and @pospielov work with a single container was to bring compreface to more users. I think only a few users would install docker.compose on the command line.

**Comment by bigbangus** [comment]
> @bigbangus: the reason for this ticket and @pospielov work with a single container was to bring compreface to more users. I think only a few users would install docker.compose on the command line.

Yes 100% agree. It should be in a single container to reach a greater audience. I was just trying to show it works with binding when using docker compose so why can't it work that way with templates in Unraid. What's the actual problem?

**Comment by pospielov** [comment]
@corgan2222 Please contact Unraid team, ask what workaround they see in this case. I think this is not the first time somebody face this problem.
If they won't give any solution, then I'll try to come up with something in CompreFace image

**Comment by jakowenko** [comment]
Thank you so much @pospielov for getting this single container version working.

With the help of @bentasker and @alexdelprete we got the [Home Assistant add-on](https://github.com/jakowenko/double-take-hassio-addons/tree/master/compreface) working with persistent storage! I'm going to let the HA community know as well which should help drive traffic towards CompreFace.


<img width="450" alt="Screen Shot 2021-11-27 at 1 29 24 PM" src="https://user-images.githubusercontent.com/1081811/143712678-250b73ff-2ca6-48da-b74e-9883df7c060e.png">

**Comment by alexdelprete** [comment]
> Thank you so much @pospielov for getting this single container version working.
> 
> With the help of @bentasker and @alexdelprete we got the [Home Assistant add-on](https://github.com/jakowenko/double-take-hassio-addons/tree/master/compreface) working with persistent storage! I'm going to let the HA community know as well which should help drive traffic towards CompreFace.
> 

One month ago I told you we needed the CompreFace addon so users could have the best solution (I hate DeepStack). I started this thread, and here we are...sometimes magic happens. Thanks to @pospielov and @bentasker. :)

**Comment by alexdelprete** [comment]
@pospielov a question: the addon is working fine, but some users let us notice that the addon is allocating 3-4GB of RAM at startup without releasing it. @bentasker that helped us creating the addon confirmed it's a java config setting. I checked in the container info and found this:

here's the allocated ram:

![image](https://user-images.githubusercontent.com/7027842/144069531-a4e05cfb-15cd-42d7-8828-6f4eb8efb8ac.png)

and we found these configuration env. variables:
![image](https://user-images.githubusercontent.com/7027842/144069288-25fb9834-676b-42e2-8e74-0c48b4c5ab71.png)

we were wondering if we could modify `API_JAVA_OPTS` with a lower value than `-Xmx4g`.

please take into account that on average Home Assistant users are not running on very powerful systems with tons of RAM, so this might be a problem for an 8GB system with other apps/addons running. I hope we can fine tune the memory requirement.

Thanks for any suggestion. Tagging also @jakowenko so we're all aligned.

**Comment by pospielov** [comment]
@jakowenko I have one favor to ask, could you rename CompreFace to Exadel CompreFace in your addon?

**Comment by pospielov** [comment]
@alexdelprete Java caches in memory embedding for each subject example. In my experience, 50 000 examples need about 8Gb of RAM. So I think you can reduce this number. Java consumes all the memory it sees, but garbage collector is quite effective and we tested CompreFace for memory leaks, looks like everything ok.
Also, working with mobilenet custom build will help as well, as it works with vectors of 128 numbers, not 512.

**Comment by pospielov** [comment]
@corgan2222 Any news from Unraid?

**Comment by jakowenko** [comment]
> @jakowenko I have one favor to ask, could you rename CompreFace to Exadel CompreFace in your addon?

Of course, do you want anything adjusted with the description?

"Exadel CompreFace is a leading free and open-source face recognition system."

**Comment by alexdelprete** [comment]
> So I think you can reduce this number. Java consumes all the memory it sees, but garbage collector is quite effective and we tested CompreFace for memory leaks, looks like everything ok.

Do you think 2GB is an appropriate number? We can just set the env variable at startup, right? This is good news @jakowenko.





> Also, working with mobilenet custom build will help as well, as it works with vectors of 128 numbers, not 512.

Would this simply require pulling a different image from your repo in addon's Dockerfile, right?

Thanks a lot.

**Comment by pospielov** [comment]
@jakowenko Looks good for now :)

**Comment by pospielov** [comment]
@alexdelprete I think 1Gb will be enough in most cases.
yes, running mobilenet is very easy, the only problem it supports only devices with AVX2.
To run CompreFace with mobilenet, use this command:
`docker run -it --name=CompreFace -e "API_JAVA_OPTS=-Xmx1g" -v compreface-db:/var/lib/postgresql/data -p 8000:80 exadel/compreface:0.6.1-mobilenet`
Another warning about mobilenet - all existing `face recognition` services will stop working. You can try to migrate them though: 
https://github.com/exadel-inc/CompreFace/blob/master/docs/Face-data-migration.md

**Comment by jakowenko** [comment]
> @jakowenko Looks good for now :)

Updated! Thank you again.

![Screen Shot 2021-12-01 at 11 36 55 AM](https://user-images.githubusercontent.com/1081811/144274915-9b646d1a-a86b-4ef1-8534-9a3dc0734694.png)

**Comment by alexdelprete** [comment]
> I think 1Gb will be enough in most cases.

great news! :)



> To run CompreFace with mobilenet, use this command:
> `docker run -it --name=CompreFace -e "API_JAVA_OPTS=-Xmx1g" -v compreface-db:/var/lib/postgresql/data -p 8000:80 exadel/compreface:0.6.1-mobilenet`

So we just need to adapt small things in our configuration.

@jakowenko do you confirm? or you need other info? Regarding mobilenet, I don't know if we actually need to move to that version...also because we need to migrate data if we do. Check next question

> Another warning about mobilenet - all existing `face recognition` services will stop working. You can try to migrate them though

I'd prefer to avoid this. So a question: if we reduce the memory requirement to 1GB do we really need to switch to mobilenet? I guess it is less resource-hungry than the default version, but how much resources can we actually save? when do you recommend mobilenet vs normal version?

Thanks a lot for the support.

**Comment by funzie19** [comment]
> I'd prefer to avoid this. So a question: if we reduce the memory requirement to 1GB do we really need to switch to mobilenet? I guess it is less resource-hungry than the default version, but how much resources can we actually save? when do you recommend mobilenet vs normal version?

I think, at least for Home Assistant giving the option for which container to use would be the best way to go. Mobilenet, FaceNet, and so on. Because of the wide range of hardware users run Home Assistant in. For example, lowering the ram requirement to 1GB would be great for a RaspberryPi user, but would not work because(as far as I know) it doesn't support AVX2 CPU instructions. Combination of detector plus memory configuration would be ideal in the Home Assistant setup.

**Comment by alexdelprete** [comment]
> For example, lowering the ram requirement to 1GB would be great for a RaspberryPi user, but would not work because(as far as I know) it doesn't support AVX2 CPU instructions.

AVX and AVX2 are not supported by ARM processors. CompreFace requires x86. Details [here](https://github.com/exadel-inc/CompreFace/blob/master/custom-builds/README.md).

But ven if it would, I would never recommend to use RPi to run Frigate+Double Take+Compreface to analyze high-res video streams. That's not the platform to run these things. I use several RPi 4B, but I'd never run HA on them with all the addons I use. I tried it and the performances were not acceptable, at least for my use case.

There are users with way more powerful systems than the RPi that sometimes are struggling with a couple of cameras. That's why Double Take and Frigate have proxy addons: so you can run components on other systems, adequate for the scope.

I don't even recommend RPi as a platform for HA in general...too many limitations when you run a full automation stack for the house.

**Comment by funzie19** [comment]
I haven't tried running HA on a RaspberryPi yet. I've always used a VM for it. I guess this would be best if there is a HA discussion elsewhere, but documenting and/or checking system requirements would be very useful in preventing `why isn't it working on my system?` given that users will have essentially a one-click install available now.

**Comment by alexdelprete** [comment]
Correct, but like I already wrote: CompreFace requires x86, that has AVX/AVX2 support. So one addon version if enough, for x86 users.

**Comment by pospielov** [comment]
@alexdelprete Regarding AVX2 - as far as I know, Celeron started support of AVX2 only from Tiger Lake (Q3 2020). This is one of the reasons why supporting FaceNet which requires only AVX is still important. 

MobileNet will also reduce RAM consumption for the Python part, which runs neural networks.

**Comment by alexdelprete** [comment]
> as far as I know, Celeron started support of AVX2 only from Tiger Lake (Q3 2020). This is one of the reasons why supporting FaceNet which requires only AVX is still important.

> AVX2 (also known as Haswell New Instructions) expands most integer commands to 256 bits and introduces fused multiply-accumulate (FMA) operations. They were first supported by Intel with the Haswell processor, which shipped in 2013.

I think 8y is more than enough. :)

> MobileNet will also reduce RAM consumption for the Python part, which runs neural networks.

Another argument in favour of the decision to support MobileNet. We'll use that in the addon, with 1GB java option.

**Comment by corgan2222** [comment]
> @corgan2222 Any news from Unraid?

sorry, my new job is quite demanding and I'm only at home on the weekend.
I made a [new Thread in the Forum](https://forums.unraid.net/topic/116814-named-volume-vs-bind-mounts/) and hope to get some developer Feedback on this.

**Comment by pospielov** [comment]
Thanks, I read the response on the UnRaid forum. This is the exact workaround solution I mentioned. 
I'll try to make it and publish it.

**Comment by corgan2222** [comment]
> Thanks, I read the response on the UnRaid forum. This is the exact workaround solution I mentioned. I'll try to make it and publish it.

Great to hear! Thanks a lot!

**Comment by bentasker** [comment]
@pospielov FWIW - that approach is exactly what we did to create the HA add-on, if the Postgres datadir is empty we copy in the files that are in the image (HA expects persistent storage to be in a different location, so that needed to happen anyway)

**Comment by alexdelprete** [comment]
@pospielov another solution would be to not include Postgres in the addon and let users install a separate addon that CompreFace could use. This would also eliminate the need for named volumes and the various workarounds to make it work on all platforms.

**Comment by pospielov** [comment]
@alexdelprete I don't think this is a good idea to provide a single image of CompreFace without a database. I mean it was meant to be a simple way to start it, but adding another Postgres addon is probably a more reliable solution, but definitely not simple. 
Basically, you can set up this image to work with external Postgres, all you need is to set environment variables, as described here:
https://github.com/exadel-inc/CompreFace/blob/master/docs/Architecture-and-scalability.md#postgresql
So if you set up external Postgres, an internal one still will be run, but I believe as it won't be used, it won't take many resources.

Also, the main version that should work on all platforms is the docker-compose version. I expect that a single image version should be used only if docker-compose is not an option.

**Comment by alexdelprete** [comment]
> but adding another Postgres addon is probably a more reliable solution, but definitely not simple.

the advantages I see with this idea are two:

1. bypass the named volume issue (needed only for Postgres)
2. avoid duplicating resources in case the user already runs a Postgres instance

In Home Assistant Postgres can be easily installed, as a custom component already made available via HACS, I think on Unraid (and other environments) too. We would have a lighter CompreFace single container, without any named volume issue to bypass, and an external Postgres that maybe users already have installed in their environment.

> Also, the main version that should work on all platforms is the docker-compose version. I expect that a single image version should be used only if docker-compose is not an option.

Agreed. The single image version is needed only where a component for a specific environment has to be created, and the environment doesn't support the docker-compose solution.

Moreover, if you think about it, with docker-compose you actually have separate containers, one of which is Postgres, so I'm basically proposing to merge only the CompreFace ones in a single image, and leave Postgres to the user.

**Comment by bigbangus** [comment]
I would be in favor of this strategy for Unraid as well. It's very common in Unraid to use external database containers and it makes it easier to manage the data.

**Comment by pospielov** [comment]
ok, so first of all, I released a new version that should solve the volume issue. It is released under the same tag, so you will need to pull the new version first.
Second, please try to set up CompreFace with external Postgres first: https://github.com/exadel-inc/CompreFace/blob/master/docs/Architecture-and-scalability.md#postgresql
Finally, what about giving users a choice? Will it be ok, if I add an environment variable that says should I start internal Postgres or not? 
Also, @jakowenko @corgan2222 how do you solve the volume problem in double-take for HA and UnRaid? I mean it uses sqllite and there is a volume in docker-compose of double-take

**Comment by alexdelprete** [comment]
> Second, please try to set up CompreFace with external Postgres first: https://github.com/exadel-inc/CompreFace/blob/master/docs/Architecture-and-scalability.md#postgresql

We already prepared the addon such as the user can configure these variables:

![image](https://user-images.githubusercontent.com/7027842/145072609-388a2782-0ced-4c94-8d03-edb95d71f960.png)

The problem will be migrating data for the existing users, from the internal postgres to the external one...

> Finally, what about giving users a choice? Will it be ok, if I add an environment variable that says should I start internal Postgres or not?

Yes I think this is a nice idea, we could also leverage this to migrate data: when the user switches from internal to external, we could maybe run a pg_dump / pg_dumpall. This is what I used to migrate data from an old version of CompreFace (with an old postgres) to the newest single image, which had a much updated Postgres.

The only thing left to address is how to migrate from FaceNet to MobileNet only the first time...is there a way to programatically know when to run the API call you told us when switching? Or can we hit that API everytime, even if not needed?

**Comment by pospielov** [comment]
What about adding both moving the data and migration into another dockerfile and adding it as a migration addon?
I mean moving data definitely not the functionality we need to add to CompreFace. 
> Or can we hit that API everytime, even if not needed?
Migration was created in a way if it fails, you can continue from the failing point. So basically yes - you can hit that API many times and it won't do anything bad. At least it was planned like this.

**Comment by pospielov** [comment]
yes, this is CPU intensive task as it has to run a neural network once again for every image. 
We want to add migration as a part of CompreFace, so it will be easy to run it from UI and see the progress, but it is not implemented and it won't be in the nearest release as well. 
If Postgres is updated to a newer version - then it will be another type of migration.
I expect that the user chooses the custom build somewhere at the beginning and usually doesn't need a migration, because he is just testing.

**Comment by pospielov** [comment]
I added the ability not to run internal DB. But before I push it to the official repository, want to check if it is what you wanted.
So, I added support of the new environment variable EXTERNAL_DB, if it's true - then I don't run internal Postgres.
Just an example of how I tested it:
```yaml
version: '3.4'

volumes:
  postgres-data2:

services:
  compreface-postgres-db:
    image: postgres:11.5
    restart: always
    container_name: "compreface-postgres-db"
    environment:
      - POSTGRES_USER=testuser
      - POSTGRES_PASSWORD=123
      - POSTGRES_DB=testdb
    volumes:
      - postgres-data2:/var/lib/postgresql/data

  compreface:
    image: pospielov/compreface-test
    restart: always
    container_name: "compreface"
    ports:
      - "8000:80"
    environment:
      - POSTGRES_USER=testuser
      - POSTGRES_PASSWORD=123
      - POSTGRES_URL=jdbc:postgresql://compreface-postgres-db:5432/testdb
      - EXTERNAL_DB=true
    depends_on:
      - compreface-postgres-db
```
I used docker-compose just because it's the easiest way for me to start Postgres and up it to the same network as CompreFace. 
So if you are OK with this solution, I'll run GitHub action and it will update the official CompreFace version

**Comment by alexdelprete** [comment]
> We want to add migration as a part of CompreFace, so it will be easy to run it from UI and see the progress, but it is not implemented and it won't be in the nearest release as well.

For Facenet->Mobilenet migration, I guess we can't do anything about it, you will have to take care of that. So the sooner we switch to Mobilenet the better, otherwise users of previous versions of the addon won't be able to migrate. @jakowenko 

> If Postgres is updated to a newer version - then it will be another type of migration.

The solution for me was using the sql dump (pg_dump), it worked well and it was fast.

**Comment by alexdelprete** [comment]
> So, I added support of the new environment variable EXTERNAL_DB, if it's true - then I don't run internal Postgres.

Just to be clear: this means postgres is still in, but it won't be started, correct?

> I pushed this version to my repository, so you can test it if you want.

So the image is already available and we can use it to test the addon? Great. cc @jakowenko

**Comment by pospielov** [comment]
> Just to be clear: this means postgres is still in, but it won't be started, correct?

yes, correct

**Comment by corgan2222** [comment]
Hello,

for me personally, a version with the possibility to use an existing Postgres would be great.
But for the average user, it's more work with more potential pitfalls.
So if it is possible to create a container with Postgres on persistent data, that would be my favorite. 
In the Community App Store are a lot of apps with two versions, one with Database included and one without. 

Your solution with the EXTERNAL_DB variable is great.
Does the DB get created on installation if it not exist?

I'm a little confused. Which container URL should I test on Unraid?
atm Im using exadel/compreface:0.6.1-arcface-r100-gpu

**Comment by pospielov** [comment]
When I build an image, I install Postgres and init it with the user and database.
Then, when you start CompreFace, regardless of variable EXTERNAL_DB, CompreFace connects to DB using environment variables POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_URL. Then it runs migrations scripts that initialize DB - creates tables, creates demo app, etc. 
So if you use external DB, all you need is to create manually the user and database, so you can set them in environment variables.
The only thing that EXTERNAL_DB does is just say if CompreFace should start internal Postgres. 

> I'm a little confused. Which container URL should I test on Unraid?

I pushed it to `pospielov/compreface-test`. It uses FaceNet as the default build.
Looks like everybody is OK with the current EXTERNAL_DB approach, so I started GitHub action to push a new version. 
https://github.com/exadel-inc/CompreFace/actions/runs/1569849037
As soon as it finishes you can test the official version `exadel/compreface:0.6.1-arcface-r100-gpu`

**Comment by alexdelprete** [comment]
The only problem with this approach is that the named volume issue is not resolved because postgres is still part of the addon.

My proposal to strip it from the image was to avoid having named volumes. With this approach the only benefit is not to start internal postgres in case the user already uses its own external instance of postgres.

It's better having a choice anyway, so it's an improvement.


> Looks like everybody is OK with the current EXTERNAL_DB approach, so I started GitHub action to push a new version.

I would have liked @jakowenko to give a feedback, but I don't see him around since a few days...

> As soon as it finishes you can test the official version `exadel/compreface:0.6.1-arcface-r100-gpu`

why arcface-r100-gpu?

**Comment by corgan2222** [comment]
external_db works great on unraid! Tnxs @pospielov 

![grafik](https://user-images.githubusercontent.com/12233951/145735420-5fabbac2-8de9-4d99-a597-6f7c0c9dc9d5.png)

**Comment by corgan2222** [comment]
But I had to give the user Superuser rights. Without superuser, I get these errors:

![grafik](https://user-images.githubusercontent.com/12233951/145735595-5a95ab15-57c0-4f23-9afb-1f8775c92a19.png)

**Comment by pospielov** [comment]
@alexdelprete didn't get about the named volume issue. 
If the user uses internal Postgres, he can use named volume or bind volume, it should work now in both cases.
If the user uses external Postgres, he just don't add any volumes for CompreFace

@corgan2222 some bad decision at the beginning of CompreFace... I'll create an issue, this is totally not ok to require superuser permissions. Not sure how to fix it now and not affect current users.

**Comment by corgan2222** [comment]
@pospielov Ok, thanks for the info. It's just one Detail I have to write in the description to avoid unnecessary bug reports.
Will test this version later without external Postgres and let you know if this works on unraid.

**Comment by bentasker** [comment]
@pospielov 

> If the user uses external Postgres, he just don't add any volumes for CompreFace

Docker will create an anonymous volume at each container startup, because the DockerFile contains a `VOLUME` declaration. Each of those will be the size of the postgres dir in your image (so, small but non-0)

Could you remove the volume declaration from the DockerFile? It doesn't preclude the user from specifying a volume at runtime, but there's no way to "unset" a volume that's been declared in the DF (one of Docker's odd foibles).

**Comment by corgan2222** [comment]
> @alexdelprete didn't get about the named volume issue. If the user uses internal Postgres, he can use named volume or bind volume, it should work now in both cases. If the user uses external Postgres, he just don't add any volumes for CompreFace

Works great now with internal and external DB on Unraid.
Great Job, tnxs!

Is there an easy way to get the data from the "old" docker-compose database into the standalone db?
I looked into https://github.com/exadel-inc/CompreFace/blob/master/docs/Face-data-migration.md but i don't get it.

**Comment by bigbangus** [comment]
> Works great now with internal and external DB on Unraid.

What version of postgres should we use on Unraid with compreface. 

Options available:
registry.hub.docker.com/library/postgres:10
registry.hub.docker.com/library/postgres:11
registry.hub.docker.com/library/postgres:12
postgres:12.5-alpine
registry.hub.docker.com/library/postgres:13
registry.hub.docker.com/library/postgres:14

**Comment by corgan2222** [comment]
I use postgres:11  https://lime-technology.com/forums/topic/69483-support-jj9987-postgresql/ from jj9987

**Comment by pospielov** [comment]
> Docker will create an anonymous volume at each container startup, because the DockerFile contains a VOLUME declaration. Each of those will be the size of the postgres dir in your image (so, small but non-0)

@bentasker yes you are right. I deleted the volume and run the rebuilding of the image.

**Comment by pospielov** [comment]
> Is there an easy way to get the data from the "old" docker-compose database into the standalone db?
I looked into https://github.com/exadel-inc/CompreFace/blob/master/docs/Face-data-migration.md but i don't get it.

There is no easy way. The documentation is about migrating data inside the database when you change the model. 
To move data between databases is quite a standard thing, I think this should help:
https://stackoverflow.com/questions/1237725/copying-postgresql-database-to-another-server
Just expose the Postgres port on CompreFace, so this script can connect to it directly.

**Comment by pospielov** [comment]
> What version of postgres should we use on Unraid with compreface.
I tested with Postgres 11.5 and Postgres 13.5. Both worked fine.

**Comment by alexdelprete** [comment]
> 

Thank you Pospielov, so now we have the base image without the named volume and using a variable the user can decide if he wants to use the internal postgres or the external one. In case he wants to use the internal one, a volume has to be created, right? We'll have to provide some instructions regarding this in the addon readme.

Thanks a lot for making all this possible.

**Comment by pospielov** [comment]
> In case he wants to use the internal one, a volume has to be created, right?
It can work without volume, but then the user will loose all the data if he deletes the volume

**Comment by alexdelprete** [comment]
> It can work without volume, but then the user will lose all the data if he deletes the volume

Obviously we'll recommend creating a persistent storage for data, so the creation of a volume.

**Comment by alexdelprete** [comment]
@pospielov the tag `exadel/compreface:0.6.1-mobilenet` would pull the latest official single-image version with `EXTERNAL_DB `support, from the official exadel repository, right? Because I didn't find any reference in the documentation about `EXTERNAL_DB`, etc.

So with a docker-compose like this I would have a single image version of Compreface MobileNet version, with an external Postgres (they have to manually create an empty db, and the user/pw), correct?

```yaml
version: '3.4'

  compreface:
    image: exadel/compreface:0.6.1-mobilenet
    restart: unless-stopped
    container_name: "compreface"
    ports:
      - "8000:80"
    environment:
      - POSTGRES_USER=compreface
      - POSTGRES_PASSWORD=password
      - POSTGRES_URL=jdbc:postgresql://postgres.mydomain.lan:5432/comprefacedb
      - EXTERNAL_DB=true
      - API_JAVA_OPTIONS=-Xmx1g
      - ADMIN_JAVA_OPTIONS=-Xmx1g
```

**Comment by pospielov** [comment]
yes, correct
https://github.com/exadel-inc/CompreFace/blob/master/docs/Installation-options.md#single-docker-container
here is the documentation

**Comment by alexdelprete** [comment]
Thanks for confirming. Docs are very well written.

I tried spinning up the single container version on a small debian server, a virtual machine on proxmox: I assigned 15GB of storage space, and during the startup, it went out of space. So I gave it 50GB, and restarted: it went up to 48GB. :(

Is it normal? Why is it taking up so much space? The DB was my central postgres server.

**Comment by pospielov** [comment]
It definitely shouldn't be like this... 
Unpacked `exadel/compreface:0.6.1` image takes 4.32GB of disk space according to docker, `exadel/compreface:0.6.1-mobilenet` image takes only 2.35GB of disk space.
![image](https://user-images.githubusercontent.com/3736126/149143588-92717f48-fa66-4814-8457-9e87c60be76a.png)
After starting `exadel/compreface:0.6.1-mobilenet` container, the disk usage didn't increase. However RECLAIMABLE space is decreased, so I expect that this is the data I created before, so probably it still takes about 1Gb space more.
Still, it's not even 15Gb of disk space

Could you run:
`docker system df`
`docker images`
and share results?

**Comment by alexdelprete** [comment]
Maybe @bentasker can shed some light about this...

**Comment by alexdelprete** [comment]
I digged into it a little bit, looking for the top 5 folders by size:

```
39526576	.
38836862	./var
38678736	./var/lib
38620922	./var/lib/docker
38616983	./var/lib/docker/vfs
```

The beast is vfs/dir, top 5 in there:
```
root@compreface:/var/lib/docker/vfs/dir# du -a | sort -n -r | head -n 5
38617420	.
1372400	./fda1856999f8067977a926ddf996a6e1a0662d568f5f60a5fa87ba7f7d5380da
1371829	./82d04c404aa2d8f9126d09570da5be3226230abf580cadaee13e34ac0bd3fd31
1371820	./fda1856999f8067977a926ddf996a6e1a0662d568f5f60a5fa87ba7f7d5380da-init
1371809	./74c77d64fb806da605adabe72be0b9656cca36d5d902afa7a42d82676ef83da8
```

Checked one of them:

```
root@compreface:/var/lib/docker/vfs/dir/fda1856999f8067977a926ddf996a6e1a0662d568f5f60a5fa87ba7f7d5380da# cat startup.sh
#!/bin/bash

# EXTERNAL_DB defines if we need to run internal DB
external_db=${EXTERNAL_DB:-false}
if [ "$external_db" = false ] ; then
    # restore default data if it was cleared by volume creation
    if [ -z "$(ls -A $PGDATA)" ]; then
       echo "Postgres directory is empty. Copy default values into it"
       cp -r /var/lib/postgresql/default/* $PGDATA
    fi
    # change permissions in case they were corrupted
    chown -R postgres:postgres $PGDATA
    chmod 700 $PGDATA

    echo Starting compreface-postgres-db
    supervisorctl start compreface-postgres-db
fi

# wait until DB starts
sleep 10
echo Starting compreface-admin
supervisorctl start compreface-admin

# wait until compreface-admin make all migrations
sleep 10
echo Starting compreface-api
supervisorctl start compreface-api

# wait until compreface-admin starts
sleep 10
echo Starting compreface-fe
supervisorctl start compreface-feroot@compreface:/var/lib/docker/vfs/dir/fda1856999f8067977a926ddf996a6e1a0662d568f5f60a5fa87ba7f7d5380da#
```

Checked another one, looks the same...I guess they're all the same...

```
root@compreface:/var/lib/docker/vfs/dir/74c77d64fb806da605adabe72be0b9656cca36d5d902afa7a42d82676ef83da8# cat startup.sh
#!/bin/bash

# EXTERNAL_DB defines if we need to run internal DB
external_db=${EXTERNAL_DB:-false}
if [ "$external_db" = false ] ; then
    # restore default data if it was cleared by volume creation
    if [ -z "$(ls -A $PGDATA)" ]; then
       echo "Postgres directory is empty. Copy default values into it"
       cp -r /var/lib/postgresql/default/* $PGDATA
    fi
    # change permissions in case they were corrupted
    chown -R postgres:postgres $PGDATA
    chmod 700 $PGDATA

    echo Starting compreface-postgres-db
    supervisorctl start compreface-postgres-db
fi

# wait until DB starts
sleep 10
echo Starting compreface-admin
supervisorctl start compreface-admin

# wait until compreface-admin make all migrations
sleep 10
echo Starting compreface-api
supervisorctl start compreface-api

# wait until compreface-admin starts
sleep 10
echo Starting compreface-fe
```

**Comment by pospielov** [comment]
I wanted to see what is inside this directory on my machine and I didn't find it.
I tried to google this folder and this is what I found: 
https://docs.docker.com/storage/storagedriver/vfs-driver/
> How the vfs storage driver works
> VFS does not support copy-on-write (COW), so each time a new layer is created, it is a deep copy of its parent layer. These layers are all located under /var/lib/docker/vfs/dir/.

It looks like this storage driver uses disk space in a very non-optimal way. Here is from [docker documentation](https://docs.docker.com/storage/storagedriver/select-storage-driver/):
> The vfs storage driver is intended for testing purposes, and for situations where no copy-on-write filesystem can be used. Performance of this storage driver is poor, and is not generally recommended for production use.

Not sure why it is chosen on your docker version

**Comment by bentasker** [comment]
> Maybe @bentasker can shed some light about this...

Sorry, been busy.

@pospielov is correct, this is because your docker instance is using VFS, it's quite inefficient compared to Overlay(2). VFS gets used on a variety of devices/environments though because it's robust and supported (more or less) everywhere.

But, @pospielov, the image does use a *lot* of layers, can I suggest removing some of those? In the dockerfile rather than having lots of `RUN` directives, chain commands where possible.

As a particular example
```
RUN apt-get update && apt-get install -y lsb-release
RUN sh -c 'echo "deb http://apt.postgresql.org/pub/repos/apt $(lsb_release -cs)-pgdg main" > /etc/apt/sources.list.d/pgdg.list'
RUN wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
RUN apt-get update && apt-get install -y postgresql-13 \
    && rm -rf /var/lib/apt/lists/*
```

You'll have 4 layers here, the first 3 of which will contain all of `apt`'s files (so probably a couple hundred meg).

If instead you do
```
RUN apt-get update && apt-get install -y lsb-release &&\
sh -c 'echo "deb http://apt.postgresql.org/pub/repos/apt $(lsb_release -cs)-pgdg main" > /etc/apt/sources.list.d/pgdg.list' &&\
wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add - && \
apt-get update && apt-get install -y postgresql-13 \
    && rm -rf /var/lib/apt/lists/*
```
Then you'll have a single layer, and `apt`'s files will never be included in it.

That way VFS users won't see massive storage usage.

**Comment by alexdelprete** [comment]
Hi @bentasker / @pospielov,

I just finished migrating all my physical servers to a central proxmox server. I'm using mainly LXC containers to optimize resource usage. The LXC container I used for Compreface is a clean and minimal Debian 11, headless, only openssh+docker+compose plugin.

After the container went up, I used compose to spin up the service using this:

```
version: '3.4'
services:
  compreface:
    image: exadel/compreface:0.6.1-mobilenet
    restart: unless-stopped
    container_name: "compreface"
    ports:
      - "8000:80"
    environment:
      - POSTGRES_USER=compreface
      - POSTGRES_PASSWORD=password
      - POSTGRES_URL=jdbc:postgresql://postgres.mydomain.lan:5432/comprefacedb
      - EXTERNAL_DB=true
      - API_JAVA_OPTS=-Xmx1g
      - ADMIN_JAVA_OPTS=-Xmx1g
```

As soon as it starts unpacking, I noticed the storage being consumed very very fast, and the process interrupted several times with out of disk errors. So I had to expand it 4 times, adding 5GB each time, reaching 40GB, only to complete the installation process. I didn't configure Docker specifying VFS, and the installation of docker was made with their official procedure.

I'm no docker fan nor expert, where should I check for VFS config, and modify it with overlay2? Is it something that has to be done at docker's config level or in the compose file?

Thanks for the help.

**Comment by bentasker** [comment]
It's configured at the docker level, so has a system wide effect, there are docs here - https://docs.docker.com/storage/storagedriver/overlayfs-driver/ 

It doesn't sound like there is in this case, but it's a little more a pain if you've got existing/other containers you want to preserve (as you start having to copy stuff about).

I've never tried it within a LXC container though

**Comment by alexdelprete** [comment]
I've installed a central Docker LXC container with 15 services on it, and it's occupying very little space and also memory, very happy about it. I think it doesn't make sense to redo everything again, as it depends on how the specific docker service is built. Probably those 15 I installed are pretty simple and I don't notice the effect I had with Compreface.

Strange thing is that I now started a Debian 11 virtual machine, and installed Compreface on it with no issue, because it's using overlay2 in there. Problem is that the VM consumes a lot of resources respect to the LXC container.

Thanks for all the explanations Ben, lot to learn about these things...and I'm just starting, because I never really liked Docker...maybe I'll start taking a look at Podman, they told me there's less overhead there...and it should be compatible with docker configs and compose files.

**Comment by pospielov** [comment]
@alexdelprete Looks like this is a common problem with proxmox LXC and docker:
https://forum.proxmox.com/threads/docker-in-lxc-container.45204/
https://forum.proxmox.com/threads/docker-lxc-unprivileged-container-on-proxmox-7-with-zfs.99796/
https://forum.proxmox.com/threads/another-docker-experience-on-proxmox.84770/

@bentasker This dockerfile was created in such a non-optimal way to simplify its support. It consists of parts of original dockerfiles. If I optimize it, it will be harder to update it if we update another docker files. But I believe it won't help. As alexdelprete does not build CompreFace in his servers, he just uses it.
So when any container creates or changes file, it doubles its size. It could potentially be even because of logs and server temp files that we can't control.

**Comment by alexdelprete** [comment]
> Not sure why it is chosen on your docker version

I found out that when using ZFS, and you install docker in a LXC container, docker will revert to VFS for compatibility issues with overlay2 and ZFS. :(

Now I created a VM to bypass the issue, so in there overlay2 is the default and Compreface installs correctly. :)

The only problem is that the mobilenet version didn't work correctly, gave me a lot of errors, couldn't even create a new app/service after the login. So I reverted to `0.6.1-facenet`.

**Comment by alexdelprete** [comment]
> If I optimize it, it will be harder to update it if we update another docker files

but in the single-image version, it could actually be optimized since it's not the typical one that will be built, right?

**Comment by alexdelprete** [comment]
@bentasker solved it with the fuse-overlayfs driver in the container: it's compatible with ZFS. Credit to [this ](https://c-goes.github.io/posts/proxmox-lxc-docker-fuse-overlayfs/)guide.

Now I have a container which is the main docker system, that manages 12 containers, and it's on VFS. :(

I wonder if I can change its driver and restart it without much hassle...hope I don't have to redo everything from scratch. What do you think Ben?

**Comment by leccyril** [comment]
any documentation on main branch for this case ? it is great to have only one container for people working on some functionalities and allow only one container...
great job

**Comment by pospielov** [comment]
https://github.com/exadel-inc/CompreFace/blob/master/docs/Installation-options.md#single-docker-container
Here is the documentation of how to use the single image version of CompreFace.

**Comment by LordNex** [comment]
Since the 1.10 update, I haven't been able to get this to install through Home Assistant. Last error was

```

The command '/bin/bash -c apt-get update && apt-get install jq -y && rm -rf /var/lib/apt/lists/*' returned a non-zero code: 100
```

**Comment by pospielov** [comment]
Do you build it from scratch or use the published image? 
I just tried the published image, and it works fine, at least the default one.

I also don't get at what stage you have this error, as I don't recall us installing or use anywhere `jq` tool

**Comment by Iceman248** [comment]
Has anyone used CompreFace with postgres 14?

**Comment by jjvelar** [comment]
Hi!
How can I access CompreFace login page from Nabu Casa remote UI?
Apparently I can't access port 8000 from Nabucasa remote UI.
Thanks and best regards,

Jose

**Comment by SgtBatten** [comment]
I have a reasonably stable install using the HA addon, but wanted to take the workload off that and put it on my unraid server. Starting from scratch with a clean install i am having issues where after some time (usually after interacting with deleting traned images or similar) something breaks and compreface cannot be recovered.

**Comment by pospielov** [comment]
This is the worst thing about the Single Container version. It's hard to check what is wrong. 
What is in logs?

**Comment by toddstar** [comment]
> I have a reasonably stable install using the HA addon, but wanted to take the workload off that and put it on my unraid server. Starting from scratch with a clean install i am having issues where after some time (usually after interacting with deleting traned images or similar) something breaks and compreface cannot be recovered.

If you removing it from HA then I'd just use standard install process as you're no longer limited to just 1 container
